RIOT-OS/RIOT
# boards/mpfs-icicle-kit: add initial support (RV64IMAC core)

### Contribution description

This is a draft PR which adds support for the Microchip PolarFire SoC (MPFS) Icicle kit Engineering Sample board. This board features a quad RV64GC core + one RV64IMAC (and a whole FPGA fabric). Only the single RV64IMAC (Our first 64bits MCU !) support is added by this PR as the idea is to let Linux runs on the quad RV64GC alongside RIOT on the single RV64IMAC.

You can find more informations for the board on [microchip/microsemi website](https://www.microsemi.com/existing-parts/parts/152514) and the SoC documentation is available [here](https://www.microsemi.com/product-directory/soc-fpgas/5498-polarfire-soc-fpga#documentation)

This PR also adds a basic support for 64bits RISC-V architecture and the PLIC controller.
In the current state, only UART, GPIO and timer peripherals (all w/ IRQ support) are implemented. Others peripherals should be added in followup PR.

I used the baremetal SDK as RIOT pkg to include vendor files and lowlevel drivers, except interruptions management which is done by RIOT PLIC driver. ([MPFS SDK link](https://github.com/polarfire-soc/platform))

SoftConsole environment can be downloaded [here](https://www.microsemi.com/product-directory/design-tools/4879-softconsole#downloads) but registration is required.

SoftConsole is not really needed but it provides the custom openOCD needed for the flashing part.
It will not longer be needed if openOCD support is upstreamed.

A lot of documentation is missing, but the board is an "usable shape". I'd like to use this draft PR to discuss some specific points like 64bits architecture for RIOT.

All comments and feedbacks are very welcome.

### Testing procedure

Beware that `make flash` is currently unsupported as the openOCD fork provided by SoftConsole environment (the default Eclipse environment for baremetal development provided by Microchip) currently doesn't implement the flash driver. There is an additionnal tool that can be used for flashing the eNVM but I didn't find time to look at it yet.
For now, RIOT only runs on the onboard LIM memory which is volatile. (One of the many reasons, this PR is a draft). Thus, in order to test any application, please use `make debug` for now. This command uses the combo openOCD/GDB to load the binary to the LIM memory.
It is also needed to use the openOCD provided by SoftConsole for now. (I hope everything will be upstream at some point).
Regarding the toolchain, standard riscv-unknown-elf-gcc can be used in theory but I didn't test it as I am using the riscv-unknown-elf-gcc toolchain provided by the SoftConsole envinronment.

Example make command for flashing `tests/shell` on `mpfs-icicle-kit-es` LIM memory.
`OPENOCD=/path/to/openocd/bin/openocd make BOARD=mpfs-icicle-kit-es -C tests/shell debug`

Please expect (a lot of) compilations errors with `tests/` and `examples/` applications because of "cast from pointer to integer of different size". This is because we heavily use int to cast void argument (like callback argument). I made some fixes for a few tests application but I don't think this subject should be handled by this PR. Thus, it should be fix in another set of PRs.

### Issues/PRs references

None.


Many thanks to @polarfire-soc for providing me this awesome board !
